File: 5.27/turtlepy/src/turtlepy/scripts/circle_motion.py
#!/usr/bin/env python3

import rospy
from geometry_msgs.msg import Twist


def main():
    rospy.init_node('circle_motion_controller')

    linear_speed = rospy.get_param('~linear_speed', 2.0)
    angular_speed = rospy.get_param('~angular_speed', 2.0)
    publish_rate_hz = rospy.get_param('~publish_rate', 10.0)

    if angular_speed == 0.0:
        rospy.logerr('angular_speed must be non-zero to generate circular motion')
        return

    publisher = rospy.Publisher('/turtle1/cmd_vel', Twist, queue_size=10)
    rate = rospy.Rate(publish_rate_hz)

    rospy.loginfo(
        'Publishing circular velocity to /turtle1/cmd_vel: linear=%.3f angular=%.3f',
        linear_speed,
        angular_speed,
    )

    while not rospy.is_shutdown():
        twist_msg = Twist()
        twist_msg.linear.x = linear_speed
        twist_msg.angular.z = angular_speed

        publisher.publish(twist_msg)
        rate.sleep()


if __name__ == '__main__':
    main()

File: 5.27/turtlecpp/src/turtlecpp/src/circle_motion.cpp
#include <geometry_msgs/Twist.h>
#include <ros/ros.h>

int main(int argc, char **argv)
{
  ros::init(argc, argv, "circle_motion_controller");
  ros::NodeHandle nh;
  ros::NodeHandle private_nh("~");

  double linear_speed = 2.0;
  double angular_speed = 2.0;
  double publish_rate_hz = 10.0;

  private_nh.param("linear_speed", linear_speed, linear_speed);
  private_nh.param("angular_speed", angular_speed, angular_speed);
  private_nh.param("publish_rate", publish_rate_hz, publish_rate_hz);

  if (angular_speed == 0.0) {
    ROS_ERROR("angular_speed must be non-zero to generate circular motion");
    return 1;
  }

  ros::Publisher cmd_vel_pub = nh.advertise<geometry_msgs::Twist>("/turtle1/cmd_vel", 10);
  ros::Rate rate(publish_rate_hz);

  ROS_INFO("Publishing circular velocity to /turtle1/cmd_vel: linear=%.3f angular=%.3f", linear_speed, angular_speed);

  while (ros::ok()) {
    geometry_msgs::Twist twist_msg;
    twist_msg.linear.x = linear_speed;
    twist_msg.angular.z = angular_speed;

    cmd_vel_pub.publish(twist_msg);
    ros::spinOnce();
    rate.sleep();
  }

  return 0;
}